test/integration/util/extension.go

```diff
@@ -0,0 +1,163 @@
+	StopCh             <-chan struct{}
+	TearDown           func()
+	ClientSet          clientset.Interface
+	RESTMapper         *restmapper.DeferredDiscoveryRESTMapper
+	APIExtensionClient apiextensionsclientset.Interface
+	SharedInformers    informers.SharedInformerFactory
+	MetadataInformers  metadatainformer.SharedInformerFactory
+	MetadataClient     metadata.Interface
+	DynamicClient      dynamic.Interface
+}
+
+// ExtensionSetup performs shared setup
+func ExtensionSetup(t *testing.T, result *kubeapiservertesting.TestServer, workerCount int) *ExtensionTestContext {
+	clientSet, err := clientset.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("error creating clientset: %v", err)
+	}
+
+	// Helpful stuff for testing CRD.
+	apiExtensionClient, err := apiextensionsclientset.NewForConfig(result.ClientConfig)
+	if err != nil {
+		t.Fatalf("error creating extension clientset: %v", err)
+	}
+	// CreateNewCustomResourceDefinition wants to use this namespace for verifying
+	// namespace-scoped CRD creation.
+	CreateNamespaceOrDie("aval", clientSet, t)
+
+	discoveryClient := cacheddiscovery.NewMemCacheClient(clientSet.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(discoveryClient)
+	restMapper.Reset()
+	config := *result.ClientConfig
+	metadataClient, err := metadata.NewForConfig(&config)
+	if err != nil {
+		t.Fatalf("failed to create metadataClient: %v", err)
+	}
+	dynamicClient, err := dynamic.NewForConfig(&config)
+	if err != nil {
+		t.Fatalf("failed to create dynamicClient: %v", err)
+	}
+	onError := func(error) bool { return true }
+	sharedInformers := informers.NewSharedInformerFactory(clientSet, 0)
+	metadataInformers := metadatainformer.NewSharedInformerFactoryWithOptions(metadataClient, 0, metadatainformer.WithStopOnError(onError))
+
+	// controller creation
+	stopCh := make(chan struct{})
+	tearDown := func() {
+		close(stopCh)
+		result.TearDownFn()
+	}
+
+	return &ExtensionTestContext{
+		StopCh:             stopCh,
+		TearDown:           tearDown,
+		RESTMapper:         restMapper,
+		ClientSet:          clientSet,
+		APIExtensionClient: apiExtensionClient,
+		DynamicClient:      dynamicClient,
+		SharedInformers:    sharedInformers,
+		MetadataInformers:  metadataInformers,
+		MetadataClient:     metadataClient,
+	}
+}
+
+func NewCRDInstance(definition *apiextensionsv1beta1.CustomResourceDefinition, namespace, name string) *unstructured.Unstructured {
+	return &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       definition.Spec.Names.Kind,
+			"apiVersion": definition.Spec.Group + "/" + definition.Spec.Version,
+			"metadata": map[string]interface{}{
+				"name":      name,
+				"namespace": namespace,
+			},
+		},
+	}
+}
+
+func CreateRandomCustomResourceDefinition(
+	t *testing.T, apiExtensionClient apiextensionsclientset.Interface,
+	dynamicClient dynamic.Interface,
+	namespace string,
+) (*apiextensionsv1beta1.CustomResourceDefinition, dynamic.ResourceInterface) {
```

Comment:
this formatting looks weird, could this line be merged to prior line?

Reply:
done